File: reorder_list.cpp
#include<bits/stdc++.h>
using namespace std;

struct ListNode{
    int val;
    ListNode* next;

};

class Solution {
public:
    void reorderList(ListNode* head) {
        vector<ListNode*> nodes;
        int n = 0;
        auto store_head = head;
        while (head){
            if (head != store_head) nodes.push_back(head);
            head = head->next;

        }

        for (int i = nodes.size()-1; i >=0; i++){
            cout<<store_head->val<<endl;
            auto store_node = store_head->next;

            if (store_node == nodes[i]){
                store_head = store_head->next;
                continue;

            }

            if (!store_node){
                store_head->next = nodes[i];
                store_head->next->next = nullptr;
                break;
            }
            store_head->next = nodes[i];
            store_head = store_head->next;
            store_head->next = store_node;
            store_head = store_head->next;
        }

    }
};

int main(){
    
    return 0;
}

File: reorderList.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head) -> None:
        """
        Do not return anything, modify head in-place instead.
        """

        nodes = []
        used_nodes = []
        store_head = head

        while (head):
            if (store_head != head) : nodes.append(head)
            head = head.next

        
        nodes.reverse()

        for node in nodes:
            print(store_head.val, store_head.next.val)
            store_next = store_head.next
            
            if store_next in used_nodes:
                store_head.next = None
                break
            
            if (store_next == node):
                if store_next is None:
                    break
                store_head = store_head.next
                continue

            if store_next is None:
                store_head.next = node
                store_head = store_head.next
                store_head.next = None
                break

            store_head.next = node
            store_head = store_head.next
            store_head.next = store_next
            store_head = store_head.next

            used_nodes.append(node)

        
